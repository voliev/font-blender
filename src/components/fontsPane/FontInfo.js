import React, { Fragment } from "react";
import classNames from "classnames";
import Preview from "@material-ui/icons/RemoveRedEyeRounded";
import PropTypes from "prop-types";

const FontInfo = ({
  family,
  category,
  subsets,
  variants,
  fetching,
  fontPreviewClickHandler
}) => {
  const fontFamilyStyle = {
    fontFamily: `'${family}', '${category}'`
  };

  const fontFamilyClasses = classNames({
    "family-name": true,
    fetching
  });

  return (
    <Fragment>
      <div className="family">
        <h4
          className={fontFamilyClasses}
          onClick={fontPreviewClickHandler}
          style={
            fetching
              ? {
                  fontFamily: `${category}`
                }
              : fontFamilyStyle
          }
          data-family={family}
          title="Preview font"
        >
          {family}
        </h4>
        <span className="preview-icon">
          <Preview fontSize="default" />
        </span>
      </div>
      <div className="category">{category}</div>
      <div className="charsets font_small">
        <span className="number">{subsets.length}</span> character{" "}
        {subsets.length === 1 ? "set" : "sets"}
      </div>
      <div className="styles font_small">
        <span className="number">{variants.length}</span>{" "}
        {variants.length === 1 ? "style" : "styles"}
      </div>
    </Fragment>
  );
};

FontInfo.propTypes = {
  family: PropTypes.string.isRequired,
  category: PropTypes.string.isRequired,
  variants: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  subsets: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  fetching: PropTypes.bool.isRequired,
  fontPreviewClickHandler: PropTypes.func.isRequired
};

export default FontInfo;
