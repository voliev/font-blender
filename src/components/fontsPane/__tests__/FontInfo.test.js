import React from "react";
import { mount, shallow } from "enzyme";
import FontInfo from "../FontInfo";
import visibleFonts from "../../../fixtures/visibleFonts";

const setup = (renderFn = shallow, propsOverrides = {}) => {
  const props = {
    ...visibleFonts[0],
    fontPreviewClickHandler: jest.fn(),
    ...propsOverrides
  };

  return {
    ...props,
    wrapper: renderFn(<FontInfo {...props} />)
  };
};

afterEach(() => {
  jest.clearAllMocks();
});

it("should render without crashing", () => {
  const { wrapper } = setup();

  expect(wrapper).toMatchSnapshot();
});

it("should receive corresponding props", () => {
  const {
    family,
    category,
    variants,
    subsets,
    fontPreviewClickHandler,
    wrapper
  } = setup(mount);

  expect(typeof wrapper.prop("family")).toBe("string");
  expect(wrapper.prop("family")).toBe(family);

  expect(typeof wrapper.prop("category")).toBe("string");
  expect(wrapper.prop("category")).toBe(category);

  expect(typeof wrapper.prop("variants")).toBe("object");
  expect(wrapper.prop("variants").hasOwnProperty("length")).toBeTruthy();
  expect(wrapper.prop("variants")).toEqual(variants);

  expect(typeof wrapper.prop("subsets")).toBe("object");
  expect(wrapper.prop("subsets").hasOwnProperty("length")).toBeTruthy();
  expect(wrapper.prop("subsets")).toBe(subsets);

  expect(typeof wrapper.prop("fetching")).toBe("boolean");
  expect(wrapper.prop("fetching")).toBe(false);

  expect(typeof wrapper.prop("fontPreviewClickHandler")).toBe("function");
  expect(wrapper.prop("fontPreviewClickHandler")).toBe(fontPreviewClickHandler);
});

it("should call handler on font family click", () => {
  const { fontPreviewClickHandler, wrapper } = setup(mount);

  wrapper.find(".family-name").simulate("click");

  expect(fontPreviewClickHandler).toHaveBeenCalled();
  expect(fontPreviewClickHandler).toHaveBeenCalledTimes(1);
});

it("should set corresponding class on font fetching", () => {
  const { wrapper } = setup(mount);

  expect(wrapper.find(".family-name").hasClass("fetching")).toBeFalsy();

  wrapper.setProps({ fetching: true });

  expect(wrapper.find(".family-name").hasClass("fetching")).toBeTruthy();
});

it("should set inline style depending on the current fetching value", () => {
  const { family, category, wrapper } = setup(mount);

  expect(wrapper.find(".family-name").prop("style")).toEqual({
    fontFamily: `'${family}', '${category}'`
  });

  wrapper.setProps({ fetching: true });

  expect(wrapper.find(".family-name").prop("style")).toEqual({
    fontFamily: `${category}`
  });
});

it("should set corresponding text depending on property array length", () => {
  const { subsets, variants, wrapper } = setup();

  expect(wrapper.find(".charsets").text()).toBe(
    `${subsets.length} character sets`
  );

  wrapper.setProps({ subsets: subsets.slice(0, 1) });

  expect(wrapper.find(".charsets").text()).toBe("1 character set");

  expect(wrapper.find(".styles").text()).toBe(`${variants.length} styles`);

  wrapper.setProps({ variants: variants.slice(0, 1) });

  expect(wrapper.find(".styles").text()).toBe("1 style");
});
