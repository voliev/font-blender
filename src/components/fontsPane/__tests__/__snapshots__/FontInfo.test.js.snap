// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<Fragment>
  <div
    className="family"
  >
    <h4
      className="family-name"
      data-family="Roboto"
      onClick={[MockFunction]}
      style={
        Object {
          "fontFamily": "'Roboto', 'sans-serif'",
        }
      }
      title="Preview font"
    >
      Roboto
    </h4>
    <span
      className="preview-icon"
    >
      <RemoveRedEyeRoundedIcon
        fontSize="default"
      />
    </span>
  </div>
  <div
    className="category"
  >
    sans-serif
  </div>
  <div
    className="charsets font_small"
  >
    <span
      className="number"
    >
      7
    </span>
     character
     
    sets
  </div>
  <div
    className="styles font_small"
  >
    <span
      className="number"
    >
      12
    </span>
     
    styles
  </div>
</Fragment>
`;
